### ISSUE TITLE
redirects show a 404 page

### ISSUE THREAD

**Issue by casperdcl** [iterative/dvc.org#2659]
Noticed after merging #2658 - when a link with an entry in https://github.com/iterative/dvc.org/blob/master/redirects-list.json is clicked, a 404 page shows for a brief period before the redirection happens. On a slow internet connection, it may make people think that the link is broken.

Should have a "redirecting" page instead of a 404. Or better... no page at all.

**Comment by shcheklein** [comment]
@casperdcl could you provide an example please?

**Comment by casperdcl** [comment]
If you go to https://dvc.org/doc/cml the 404 page flashes briefly.

**Comment by shcheklein** [comment]
I think this is related to the https://github.com/iterative/cml-website/pull/63/ ? It doesn't look like this is about dvc.org itself.

<img width="1439" alt="Screen Shot 2021-07-22 at 12 53 08 PM" src="https://user-images.githubusercontent.com/3659196/126701372-77f15641-ee61-44a1-91cb-65d6946454d5.png">

**Comment by shcheklein** [comment]
Yep, I see 404 in the logs. Need to research this a bit.

**Comment by rogermparent** [comment]
I think I know what's going on here, it's related to the difference between special Gatsby client-side links that should only be used on internal page links vs regular `a`s that are used for everything else. It's up to our `Link` component to infer which one should be used depending on the given `href`, but in this case the fact that an external link looks like an internal one tricks the link component into using the wrong behavior.

A quick fix is to allow a manual override of whether the link is a `Link` or `a`, though we could also make the inference behavior more robust and able to handle internal redirects to external sites.

**Comment by julieg18** [comment]
> A quick fix is to allow a manual override of whether the link is a `Link` or `a`, though we could also make the inference behavior more robust and able to handle internal redirects to external sites.

I've tested using an `a` tag, but it produced the same result. I think its a hydration issue, looking at [this code](https://github.com/iterative/dvc.org/blob/master/src/components/Link/index.tsx#L106). Also if you restart the page(at least on my end) the link looks as it should: 
![image](https://user-images.githubusercontent.com/43496356/127927892-400fe360-0ac5-4130-95df-48f39b0a5e99.png)

I can think of two possible solutions so far:
1. Update our `Link` component to render after mount. 
2. Have our `ResultLinkComponent` check if the `href` prop is a redirect somehow and have it use the `redirect()` function instead of just using a regular `a` link. From my testing, a `redirect()` function doesn't show the 404 page like a regular `a` link would.

**Comment by rogermparent** [comment]
I looked into the issue and replicated the fact that normal `a`s suffer, but also noticed that the link works completely fine when disabling javascript.

Turns out, dvc.org has the plugin `gatsby-plugin-catch-links` which uses a global onClick handler to force all links, including normal `a`s, to behave like Gatsby links that attempt CSR on routes that look internal.

I managed to stop the issue by

1. Turning the link in question into a plain `a`

2. Replacing the optionless `gatsby-plugin-catch-links` definition with:
```
{
  resolve: 'gatsby-plugin-catch-links',
  options: {
    excludePattern: /\/doc\/cml/
  }
}
```

And with that the link consistently doesn't show a 404 and a related issue I saw where the page navigated twice (such that I needed two back button clicks to get back to dvc.org) went away.

Basically, a lot of dvc.org is built on the assumption that links that look internal are internal, and aren't external redirects pretending to be internal links. We'll have to suppress multiple layers of behavior to use internal-looking links like they're external.

So we'll need something like the solution I proposed and then add all internal-looking links that we want to use external redirects for to the `gatsby-plugin-catch-links` exclusion list (or remove `gatsby-plugin-catch-links` and solve whatever issue caused its introduction in another way.

**Comment by julieg18** [comment]
Nice catch, @rogermparent! That explains the reason for regular `a` links not working correctly! 

> So we'll need something like the solution I proposed and then add all internal-looking links that we want to use external redirects for to the gatsby-plugin-catch-links exclusion list (or remove gatsby-plugin-catch-links and solve whatever issue caused its introduction in another way.

This solution seems to be the way to go if we want to be able to add redirect links on our site without any issues.

> Alternatively, we could just use the external cml.dev link in the sidebar instead of doing the whole redirect thing.

If we choose to go this way, I don't think we will have to edit the sidebar code. I think we just have to update our `Link` code to render links on mount(like I mentioned in my previous comment), then [this code](https://github.com/iterative/dvc.org/blob/master/src/components/Link/index.tsx#L106) will work correctly and turn `/doc/cml` into the external `cml.dev` link automatically.

We could also go with the second solution since its easier to implement for now to get rid of the bug, and implement the second solution later.

**Comment by casperdcl** [comment]
> we could just use the external cml.dev link in the sidebar instead of doing the whole redirect thing

We'll still need `/doc/cml/*` redirection in case people had copy-pasted & shared old links.

**Comment by julieg18** [comment]
> We'll still need /doc/cml/* redirection in case users had copy-pasted & shared old links.

 You are correct! If I'm understanding @rogermparent correctly, we aren't planning to take down the redirect logic, so old links `dvc.org/doc/cml/*` will still work no matter which solution we decide to go with 😀

**Comment by julieg18** [comment]
Looking through our commits, the `gatsby-plugin-catch-links` have been [part of the config](https://github.com/iterative/dvc.org/commit/7fafc0b09997aa6ccaf6a648fe9e667e3a7cb558#diff-b5e305780d9d473da97c61beab8bc36e5e8871b360942e4686c9b20d8c5d4cfaR52) since we migrated to Gatsby. As far as I can tell, links appear to be working the same without the plugin on my local machine. @rogermparent, do you think it's safe to take it off?

**Comment by rogermparent** [comment]
> do you think it's safe to take it off?

I would assume they were added so that we could let Gatsby run CSR on links in Markdown content which don't (and shouldn't) have concept of internal/external links and the `Link` component. However, this could be done in less heavy-handed ways like using custom components in the MD engine to replace links with our `Link` component (probably twice through both blog and docs since they use different MD rendering engines :grimacing:)

However, even if the CSR-catching functionality is "broken", the links will work perfectly fine. The worst that can break is internal links being nigh-imperceptibly slower and the transition a bit less fluid.

**Comment by rogermparent** [comment]
I think this has been fixed and we just forgot to close. I tried replicating the original issue and didn't get it. Please do reopen if I'm mistaken!
